internal/rootcoord/quota_center.go

```diff
@@ -290,40 +334,84 @@ func (q *QuotaCenter) calculateReadRates() {
+	coolOff := func(realTimeSearchRate float64, realTimeQueryRate float64, collections ...int64) {
+		for _, collection := range collections {
+			if q.currentRates[collection][internalpb.RateType_DQLSearch] != Inf && realTimeSearchRate > 0 {
+				q.currentRates[collection][internalpb.RateType_DQLSearch] = Limit(realTimeSearchRate * coolOffSpeed)
+				log.Warn("QuotaCenter cool read rates off done",
+					zap.Int64("collectionID", collection),
+					zap.Any("searchRate", q.currentRates[collection][internalpb.RateType_DQLSearch]))
+			}
+			if q.currentRates[collection][internalpb.RateType_DQLQuery] != Inf && realTimeQueryRate > 0 {
+				q.currentRates[collection][internalpb.RateType_DQLQuery] = Limit(realTimeQueryRate * coolOffSpeed)
+				log.Warn("QuotaCenter cool read rates off done",
+					zap.Int64("collectionID", collection),
+					zap.Any("queryRate", q.currentRates[collection][internalpb.RateType_DQLQuery]))
+			}
 		}
-		q.guaranteeMinRate(Params.QuotaConfig.DQLMinSearchRate.GetAsFloat(), internalpb.RateType_DQLSearch)
-		q.guaranteeMinRate(Params.QuotaConfig.DQLMinQueryRate.GetAsFloat(), internalpb.RateType_DQLQuery)
-		log.Warn("QuotaCenter cool read rates off done",
-			zap.Any("searchRate", q.currentRates[internalpb.RateType_DQLSearch]),
-			zap.Any("queryRate", q.currentRates[internalpb.RateType_DQLQuery]))
-		log.Info("QueryNodeMetrics when cool-off", zap.Any("metrics", q.queryNodeMetrics))
+
+		q.guaranteeMinRate(Params.QuotaConfig.DQLMinSearchRate.GetAsFloat(), internalpb.RateType_DQLSearch, collections...)
+		q.guaranteeMinRate(Params.QuotaConfig.DQLMinQueryRate.GetAsFloat(), internalpb.RateType_DQLQuery, collections...)
+		log.Info("QueryNodeMetrics when cool-off",
+			zap.Any("metrics", q.queryNodeMetrics))
 	}
 
 	// TODO: unify search and query?
 	realTimeSearchRate := q.getRealTimeRate(internalpb.RateType_DQLSearch)
 	realTimeQueryRate := q.getRealTimeRate(internalpb.RateType_DQLQuery)
 
-	queueLatencyFactor := q.getQueryLatencyFactor()
-	if Limit(queueLatencyFactor) == Limit(coolOffSpeed) {
-		coolOff(realTimeSearchRate, realTimeQueryRate)
+	enableQueueProtection := Params.QuotaConfig.QueueProtectionEnabled.GetAsBool()
+	if !enableQueueProtection {
 		return
 	}
 
-	queueLengthFactor := q.getNQInQueryFactor()
-	if Limit(queueLengthFactor) == Limit(coolOffSpeed) {
-		coolOff(realTimeSearchRate, realTimeQueryRate)
-		return
+	// query latency
+	queueLatencyThreshold := Params.QuotaConfig.QueueLatencyThreshold.GetAsDuration(time.Second)
+	// queueLatencyThreshold >= 0 means enable queue latency protection
+	if queueLatencyThreshold >= 0 {
+		for _, metric := range q.queryNodeMetrics {
+			searchLatency := metric.SearchQueue.AvgQueueDuration
+			queryLatency := metric.QueryQueue.AvgQueueDuration
+			if searchLatency >= queueLatencyThreshold || queryLatency >= queueLatencyThreshold {
+				collections := metric.Effect.CollectionIDs
+				coolOff(realTimeSearchRate, realTimeQueryRate, collections...)
```

Comment:
seems the collection will cool-off many times if it is loaded in many query nodes?